File: MauloaDemo.Models/Customer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Configuration;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using Newtonsoft.Json;
using CBAF;
using CBAF.Attributes;

namespace MauloaDemo.Models {

    [Table("customer")]
    public class Customer : IValidatableObject {

        [Key(), Hankaku()]
        public string c_num { get; set; }

        [Hankaku(), UpperCase()]
        public string agent_cd { get; set; }

        [Required(), Hankaku(), UpperCase()]
        public string sub_agent_cd { get; set; }


        [Required(), Hankaku(), UpperCase()]
        public string g_last { get; set; }

        [Hankaku(), UpperCase()]
        public string g_first { get; set; }

        [Hankaku(), UpperCase()]
        public string b_last { get; set; }

        [Hankaku(), UpperCase()]
        public string b_first { get; set; }

        public string g_last_kana { get; set; }
        public string g_first_kana { get; set; }
        public string b_last_kana { get; set; }
        public string b_first_kana { get; set; }
        public string g_last_kanji { get; set; }
        public string g_first_kanji { get; set; }
        public string b_last_kanji { get; set; }
        public string b_first_kanji { get; set; }

        [Required(), StringLength(3), Hankaku(), UpperCase()]
        public string area_cd { get; set; }

        [StringLength(20), Hankaku()]
        public string tour_cd { get; set; }

        [Hankaku(), UpperCase()]
        public string church_cd { get; set; }

        [DataType(DataType.Date), DisplayFormat(DataFormatString = "MM/dd/yyyy")]
        public DateTime? wed_date { get; set; }

        [DataType(DataType.Time), DisplayFormat(DataFormatString = "HH:mm")]
        public DateTime? wed_time { get; set; }

        [DataType(DataType.Time), DisplayFormat(DataFormatString = "HH:mm")]
        public DateTime? htl_pick { get; set; }

        [DataType(DataType.Date), DisplayFormat(DataFormatString = "MM/dd/yyyy")]
        public DateTime? bf_date { get; set; }

        [DataType(DataType.Time), DisplayFormat(DataFormatString = "HH:mm")]
        public DateTime? bf_time { get; set; }

        public string bf_place { get; set; }

        [DataType(DataType.Date), DisplayFormat(DataFormatString = "MM/dd/yyyy")]
        public DateTime? ft_date { get; set; }

        [DataType(DataType.Time), DisplayFormat(DataFormatString = "HH:mm")]
        public DateTime? ft_time { get; set; }

        public string ft_place { get; set; }

        [UpperCase]
        public string in_flight { get; set; }

        [UpperCase]
        public string in_dep { get; set; }

        [DataType(DataType.Date), DisplayFormat(DataFormatString = "MM/dd/yyyy")]
        public DateTime? in_dep_date { get; set; }

        [DataType(DataType.Time), DisplayFormat(DataFormatString = "HH:mm")]
        public DateTime? in_dep_time { get; set; }

        [UpperCase]
        public string in_arr { get; set; }

        [DataType(DataType.Date), DisplayFormat(DataFormatString = "MM/dd/yyyy")]
        public DateTime? in_arr_date { get; set; }

        [DataType(DataType.Time), DisplayFormat(DataFormatString = "HH:mm")]
        public DateTime? in_arr_time { get; set; }

        [UpperCase]
        public string out_flight { get; set; }

        [UpperCase]
        public string out_dep { get; set; }

        [DataType(DataType.Date), DisplayFormat(DataFormatString = "MM/dd/yyyy")]
        public DateTime? out_dep_date { get; set; }

        [DataType(DataType.Time), DisplayFormat(DataFormatString = "HH:mm")]
        public DateTime? out_dep_time { get; set; }

        [UpperCase]
        public string out_arr { get; set; }

        [DataType(DataType.Date), DisplayFormat(DataFormatString = "MM/dd/yyyy")]
        public DateTime? out_arr_date { get; set; }

        [DataType(DataType.Time), DisplayFormat(DataFormatString = "HH:mm")]
        public DateTime? out_arr_time { get; set; }

        [Hankaku(), UpperCase()]
        public string hotel_cd { get; set; }

        [NotMapped]
        public string hotel_name { get; set; }

        public string room_number { get; set; }

        [DataType(DataType.Date), DisplayFormat(DataFormatString = "MM/dd/yyyy")]
        public DateTime? checkin_date { get; set; }

        [DataType(DataType.Date), DisplayFormat(DataFormatString = "MM/dd/yyyy")]
        public DateTime? checkout_date { get; set; }

        public string note { get; set; }

        public string staff_note { get; set; }

        public short attend_count { get; set; }

        [StringLength(100)]
        public string attend_name { get; set; }

        public string attend_memo { get; set; }

        public bool cust_cxl { get; set; }

        [DataType(DataType.Date), DisplayFormat(DataFormatString = "MM/dd/yyyy")]
        public DateTime? cxl_date { get; set; }

        [Hankaku(), LowerCase()]
        public string cxl_by { get; set; }

        [DataType(DataType.DateTime), 
         DisplayFormat(DataFormatString = "MM/dd/yyyy HH:mm:ss"), 
         DatabaseGenerated(DatabaseGeneratedOption.Computed)]
        [IgnoreChangeDiff]
        public DateTime create_date { get; set; }

        [IgnoreChangeDiff]
        public string create_by { get; set; }

        [DataType(DataType.Date), DisplayFormat(DataFormatString = "MM/dd/yyyy HH:mm:ss")]
        public DateTime? final_date { get; set; }

        [Hankaku(), LowerCase()]
        public string final_by { get; set; }

        [Required(), Hankaku(), LowerCase()]
        [IgnoreChangeDiff]
        public string last_person { get; set; }

        [DataType(DataType.DateTime), 
         DisplayFormat(DataFormatString = "MM/dd/yyyy HH:mm:ss"),
         DatabaseGenerated(DatabaseGeneratedOption.Computed)]
        [IgnoreChangeDiff]
        public DateTime update_date { get; set; }

        /// 排他制御のための更新日時比較で使うための値。ユーザーのTime Zoneによる補正処理からは除外される。
        [NotMapped]
        [IgnoreChangeDiff]
        public DateTime update_date_stamp { get; set; }


        [NotMapped]
        public bool is_sunset { get; set; }

        [NotMapped]
        public bool noDupCheck { get; set; }


        [NotMapped, JsonIgnore]
        public string GroomName {
            get {
                return TypeHelper.GetStrTrim(this.g_last) 
                    + (string.IsNullOrEmpty(this.g_first) ? "" : " " + TypeHelper.GetStrTrim(this.g_first));
            }
        }

        [NotMapped, JsonIgnore]
        public string GroomNameKanji {
            get {
                return TypeHelper.GetStrTrim(this.g_last_kanji) 
                    + (string.IsNullOrEmpty(this.g_first_kanji) ? "" : " " + TypeHelper.GetStrTrim(this.g_first_kanji));
            }
        }

        [NotMapped, JsonIgnore]
        public string BrideName {
            get {
                return TypeHelper.GetStrTrim(this.b_last) 
                    + (string.IsNullOrEmpty(this.b_first) ? "" : " " + TypeHelper.GetStrTrim(this.b_first));
            }
        }

        [NotMapped, JsonIgnore]
        public string BrideNameKanji {
            get {
                return TypeHelper.GetStrTrim(this.b_last_kanji)
                    + (string.IsNullOrEmpty(this.b_first_kanji) ? "" : " " + TypeHelper.GetStrTrim(this.b_first_kanji));
            }
        }

        [NotMapped, JsonIgnore]
        public string CustomerNames {
            get {
                return this.GroomName 
                    + (string.IsNullOrEmpty(this.BrideName) ? "" : " / " + this.BrideName);
            }
        }

        [NotMapped, JsonIgnore]
        public string CustomerNamesKanji {
            get {
                return this.GroomNameKanji 
                    + (string.IsNullOrEmpty(this.BrideNameKanji) ? "" : " / " + this.BrideNameKanji);
            }
        }


        [NotMapped]
        public string wed_time_s {
            get {
                if (is_sunset) {
                    return "Sunset";
                } else { 
                    return TypeHelper.TimeHHmm(this.wed_time);
                }
            }
            set {
                var SunsetBlockTime = TypeHelper.GetStrTrim(ConfigurationManager.AppSettings["SunsetBlockTime"]);
                if (string.Equals(value, "Sunset")) {
                    value = SunsetBlockTime;
                }
                this.wed_time = TypeHelper.GetDateTimeOrNull(value);
            }
        }

        [NotMapped]
        public string htl_pick_s {
            get {
                return TypeHelper.TimeHHmm(this.htl_pick);
            }
            set {
                this.htl_pick = TypeHelper.GetDateTimeOrNull(value);
            }
        }

        [NotMapped]
        public string bf_time_s {
            get {
                return TypeHelper.TimeHHmm(this.bf_time);
            }
            set {
                this.bf_time = TypeHelper.GetDateTimeOrNull(value);
            }
        }

        [NotMapped]
        public string ft_time_s {
            get {
                return TypeHelper.TimeHHmm(this.ft_time);
            }
            set {
                this.ft_time = TypeHelper.GetDateTimeOrNull(value);
            }
        }

        [NotMapped]
        public string in_dep_time_s {
            get {
                return TypeHelper.TimeHHmm(this.in_dep_time);
            }
            set {
                this.in_dep_time = TypeHelper.GetDateTimeOrNull(value);
            }
        }

        [NotMapped]
        public string in_arr_time_s {
            get {
                return TypeHelper.TimeHHmm(this.in_arr_time);
            }
            set {
                this.in_arr_time = TypeHelper.GetDateTimeOrNull(value);
            }
        }

        [NotMapped]
        public string out_dep_time_s {
            get {
                return TypeHelper.TimeHHmm(this.out_dep_time);
            }
            set {
                this.out_dep_time = TypeHelper.GetDateTimeOrNull(value);
            }
        }

        [NotMapped]
        public string out_arr_time_s {
            get {
                return TypeHelper.TimeHHmm(this.out_arr_time);
            }
            set {
                this.out_arr_time = TypeHelper.GetDateTimeOrNull(value);
            }
        }

        public class ChangedRows {
            public ICollection<string> c_num { get; set; }
            public ICollection<bool> dup_check_done { get; set; }
        }

        public class Note {
            public string c_num { get; set; }
            public string note { get; set; }
            public bool edit_all { get; set; }
        }

        public void ValidateSave() {
            if (string.IsNullOrEmpty(this.g_last)) {
                throw new Exception("Please input a last name for the groom.");
            }
            if (string.IsNullOrEmpty(this.area_cd)) {
                throw new Exception("Please select an area.");
            }
            if (string.IsNullOrEmpty(this.sub_agent_cd)) {
                throw new Exception("Please select an agent code.");
            }
            if (string.IsNullOrEmpty(this.last_person)) {
                throw new Exception("User id is required in order to save customer.");
            }
        }


        public IEnumerable<ValidationResult> Validate(ValidationContext validationContext) {
            if (!string.IsNullOrEmpty(this.c_num) && !HankakuZenkaku.isHankakuAlphaNumericOrSymbol(this.c_num, false)) {
                yield return new ValidationResult(string.Format("c_num は半角英数字で入力してください。({0})", this.c_num), new[] { "c_num" });
            }

            if (!string.IsNullOrEmpty(this.agent_cd) && !HankakuZenkaku.isHankakuAlphaNumericOrSymbol(this.agent_cd, false)) {
                yield return new ValidationResult(string.Format("agent_cd は半角英数字で入力してください。({0})", this.agent_cd), new[] { "agent_cd" });
            }

            if (!string.IsNullOrEmpty(this.sub_agent_cd) && !HankakuZenkaku.isHankakuAlphaNumericOrSymbol(this.sub_agent_cd, false)) {
                yield return new ValidationResult(string.Format("sub_agent_cd は半角英数字で入力してください。({0})", this.sub_agent_cd), new[] { "sub_agent_cd" });
            }

            if (!string.IsNullOrEmpty(this.g_last) && !HankakuZenkaku.isHankakuAlphaNumericOrSymbol(this.g_last, true)) {
                yield return new ValidationResult(string.Format("g_last は半角英数字で入力してください。({0})", this.g_last), new[] { "g_last" });
            }

            if (!string.IsNullOrEmpty(this.g_first) && !HankakuZenkaku.isHankakuAlphaNumericOrSymbol(this.g_first, true)) {
                yield return new ValidationResult(string.Format("g_first は半角英数字で入力してください。({0})", this.g_first), new[] { "g_first" });
            }

            if (!string.IsNullOrEmpty(this.b_last) && !HankakuZenkaku.isHankakuAlphaNumericOrSymbol(this.b_last, true)) {
                yield return new ValidationResult(string.Format("b_last は半角英数字で入力してください。({0})", this.b_last), new[] { "b_last" });
            }

            if (!string.IsNullOrEmpty(this.b_first) && !HankakuZenkaku.isHankakuAlphaNumericOrSymbol(this.b_first, true)) {
                yield return new ValidationResult(string.Format("b_first は半角英数字で入力してください。({0})", this.b_first), new[] { "b_first" });
            }
        }
    }
}

File: MauloaDemo.Web/Scripts/App/models/customer.ts
declare class Customer extends kendo.data.Model {
    constructor(options: any);
    parseJSON(): void;
    toJSON(): any;
    GroomName(): string;
    BrideName(): string;
    getWedDateStr(): string;
    save(options);
    static fetch(c_num, options);
    static updateFinalInfo(c_num, finalize, options);
}

define(function () {

    //*-------------------------------- Customer Model　--------------------------------*//
    var Customer = <typeof Customer>kendo.data.Model.define({
        id: "c_num",
        fields: {
            c_num: { type: "string", defaultValue: "" }

            , g_last: { type: "string", validation: { required: true } }
            , g_first: { type: "string" }
            , b_last: { type: "string" }
            , b_first: { type: "string" }

            , g_last_kana: { type: "string" }
            , g_first_kana: { type: "string" }
            , b_last_kana: { type: "string" }
            , b_first_kana: { type: "string" }

            , g_last_kanji: { type: "string" }
            , g_first_kanji: { type: "string" }
            , b_last_kanji: { type: "string" }
            , b_first_kanji: { type: "string" }

            , agent_cd: { type: "string", validation: { required: true } }
            , sub_agent_cd: { type: "string", validation: { required: true } }
            , area_cd: { type: "string", validation: { required: true } }
            , tour_cd: { type: "string" }
            , church_cd: { type: "string" }

            , wed_date: { type: "date", nullable: true, defaultValue: null }
            , wed_time: { type: "date", nullable: true, defaultValue: null }
            , wed_time_s: { type: "string" }
            , htl_pick: { type: "date", nullable: true, defaultValue: null }
            , htl_pick_s: { type: "string" }

            , bf_date: { type: "date", nullable: true, defaultValue: null }
            , bf_time: { type: "date", nullable: true, defaultValue: null }
            , bf_time_s: { type: "string" }
            , bf_place: { type: "string" }

            , ft_date: { type: "date", nullable: true, defaultValue: null }
            , ft_time: { type: "date", nullable: true, defaultValue: null }
            , ft_time_s: { type: "string" }
            , ft_place: { type: "string" }

            , in_flight: { type: "string" }
            , in_dep: { type: "string" }
            , in_dep_date: { type: "date", nullable: true, defaultValue: null }
            , in_dep_time: { type: "date", nullable: true, defaultValue: null }
            , in_dep_time_s: { type: "string" }
            , in_arr: { type: "string" }
            , in_arr_date: { type: "date", nullable: true, defaultValue: null }
            , in_arr_time: { type: "date", nullable: true, defaultValue: null }
            , in_arr_time_s: { type: "string" }

            , out_flight: { type: "string" }
            , out_dep: { type: "string" }
            , out_dep_date: { type: "date", nullable: true, defaultValue: null }
            , out_dep_time: { type: "date", nullable: true, defaultValue: null }
            , out_dep_time_s: { type: "string" }
            , out_arr: { type: "string" }
            , out_arr_date: { type: "date", nullable: true, defaultValue: null }
            , out_arr_time: { type: "date", nullable: true, defaultValue: null }
            , out_arr_time_s: { type: "string" }

            , hotel_cd: { type: "string" }
            , room_number: { type: "string" }
            , checkin_date: { type: "date", nullable: true, defaultValue: null }
            , checkout_date: { type: "date", nullable: true, defaultValue: null }
            , note: { type: "string" }
            , staff_note: { type: "string" }
            , attend_count: { type: "number", defaultValue: 0 }
            , attend_name: { type: "string" }
            , attend_memo: { type: "string" }

            , cust_cxl: { type: "boolean", defaultValue: false }
            , cxl_date: { type: "date", nullable: true, defaultValue: null }
            , cxl_date_s: { type: "string" }
            , cxl_by: { type: "string" }

            , create_date: { type: "date", defaultValue: null }
            , create_date_s: { type: "string" }
            , create_by: { type: "string" }
            , final_date: { type: "date", nullable: true, defaultValue: null }
            , final_by: { type: "string" }
            , update_date: { type: "date", defaultValue: null }
            , update_date_s: { type: "string" }
            , update_date_stamp: { type: "date", defaultValue: null }
        }
    });

    Customer.prototype.GroomName = function(){
        return this.get("g_first") + " " + this.get("g_last");
    };

    Customer.prototype.BrideName = function() {
        return this.get("b_first") + " " + this.get("b_last");
    };

    Customer.prototype.parseJSON = function() {
        this.set("id", this.get(this.idField)); //idを明示的にセット。isNew()の値を正しく取得するために必要。
        this.wed_date = App.Utils.convertToDateTime(this.wed_date);
        this.wed_time = App.Utils.convertToDateTime(this.wed_time)
        this.wed_time_s = kendo.toString(this.wed_time, "HH:mm");
        if (this.is_sunset) {
            this.wed_time_s = "Sunset";
        }
        this.htl_pick = App.Utils.convertToDateTime(this.htl_pick)
        this.htl_pick_s = kendo.toString(this.htl_pick, "HH:mm");
        this.bf_date = App.Utils.convertToDateTime(this.bf_date);
        this.bf_time = App.Utils.convertToDateTime(this.bf_time)
        this.bf_time_s = kendo.toString(this.bf_time, "HH:mm");
        this.ft_date = App.Utils.convertToDateTime(this.ft_date);
        this.ft_time = App.Utils.convertToDateTime(this.ft_time)
        this.ft_time_s = kendo.toString(this.ft_time, "HH:mm");
        this.in_dep_date = App.Utils.convertToDateTime(this.in_dep_date);
        this.in_dep_time = App.Utils.convertToDateTime(this.in_dep_time)
        this.in_dep_time_s = kendo.toString(this.in_dep_time, "HH:mm");
        this.in_arr_date = App.Utils.convertToDateTime(this.in_arr_date);
        this.in_arr_time = App.Utils.convertToDateTime(this.in_arr_time)
        this.in_arr_time_s = kendo.toString(this.in_arr_time, "HH:mm");
        this.out_dep_date = App.Utils.convertToDateTime(this.out_dep_date);
        this.out_dep_time = App.Utils.convertToDateTime(this.out_dep_time)
        this.out_dep_time_s = kendo.toString(this.out_dep_time, "HH:mm");
        this.out_arr_date = App.Utils.convertToDateTime(this.out_arr_date);
        this.out_arr_time = App.Utils.convertToDateTime(this.out_arr_time)
        this.out_arr_time_s = kendo.toString(this.out_arr_time, "HH:mm");
        this.checkin_date = App.Utils.convertToDateTime(this.checkin_date);
        this.checkout_date = App.Utils.convertToDateTime(this.checkout_date);
        this.cxl_date = App.Utils.convertToDateTime(this.cxl_date);
        this.create_date = App.Utils.convertToDateTime(this.create_date);
        this.final_date = App.Utils.convertToDateTime(this.final_date);
        this.update_date = App.Utils.convertToDateTime(this.update_date);
        this.update_date_stamp = App.Utils.convertToDateTime(this.update_date_stamp);

        this.cxl_date_s = kendo.toString(this.cxl_date, "M/dd HH:mm:ss");
        this.create_date_s = kendo.toString(this.create_date, "M/dd HH:mm:ss");
        this.update_date_s = kendo.toString(this.update_date, "M/dd HH:mm:ss");

        this.dirty = false;
    };

    Customer.prototype.toJSON = function(){
        var json = kendo.data.ObservableObject.prototype.toJSON.call(this);
        delete json.create_date;
        delete json.cxl_date_s;
        delete json.create_date_s;
        delete json.update_date;
        delete json.update_date_s;
        delete json.dirty;
        delete json.id;

        json.wed_date = kendo.toString(this.wed_date, "yyyy/MM/dd");
        if (this.is_sunset) {
            json.wed_time_s = App.Config.SunsetBlockTime;
        }
        json.wed_time = this.wed_time_s;
        json.htl_pick = this.htl_pick_s;
        json.bf_date = kendo.toString(this.bf_date, "yyyy/MM/dd");
        json.bf_time = this.bf_time_s;
        json.ft_date = kendo.toString(this.ft_date, "yyyy/MM/dd");
        json.ft_time = this.ft_time_s;
        json.in_dep_date = kendo.toString(this.in_dep_date, "yyyy/MM/dd");
        json.in_dep_time = this.in_dep_time_s;
        json.in_arr_date = kendo.toString(this.in_arr_date, "yyyy/MM/dd");
        json.in_arr_time = this.in_arr_time_s;
        json.out_dep_date = kendo.toString(this.out_dep_date, "yyyy/MM/dd");
        json.out_dep_time = this.out_dep_time_s;
        json.out_arr_date = kendo.toString(this.out_arr_date, "yyyy/MM/dd");
        json.out_arr_time = this.out_arr_time_s;
        json.checkin_date = kendo.toString(this.checkin_date, "yyyy/MM/dd");
        json.checkout_date = kendo.toString(this.checkout_date, "yyyy/MM/dd");
        json.final_date = kendo.toString(this.final_date, "yyyy/MM/dd HH:mm:ss");
        json.update_date_stamp = kendo.toString(this.update_date_stamp, "yyyy/MM/dd HH:mm:ss");
        return json;
    };

    Customer.prototype.getWedDateStr = function(){
        var wd = this.get("wed_date");
        return kendo.toString(wd, "M/dd/yyyy");
    };

    Customer.prototype.save = function(options){
        var data = this.toJSON();
        data.noDupCheck = options.noDupCheck;
        var dataStr = JSON.stringify(data);
        //console.log("Customer.save: dataStr = " + dataStr);

        $.ajax({
            url: App.getApiPath("Customers"),
            data: dataStr,
            type: "POST",
            processData: false,
            contentType: "application/json; charset=utf-8"
        })
            .done(function (result) {
                if (options && options.success) options.success(result);
            }).fail(function (jqXHR, textStatus, errorThrown) {
                if (options && options.fail) options.fail(jqXHR, textStatus, errorThrown);
            }).always(function () {
                if (options && options.always) options.always();
            });
    };

    Customer.prototype.hasBriefingInfo = function(){
        return this.get("bf_date") && this.get("ft_date");
    };

    Customer.prototype.hasInFlightInfo = function(){
        return this.get("in_arr_date") && this.get("in_arr_time_s");
    };

    Customer.prototype.hasOutFlightInfo = function(){
        return this.get("out_dep_date") && this.get("out_dep_time_s");
    };

    Customer.prototype.hasHotelInfo = function(){
        return this.get("hotel_cd");
    };

    Customer.prototype.hasAttendantInfo = function(){
        return this.get("attend_count") || this.get("attend_name") || this.get("attend_memo");
    };

    Customer.prototype.hasAddressInfo = function(){
        var list = this.get("AddressInfos");
        return list && list.length;
    };

    Customer.prototype.hasCostumeInfo = function(){
        var list = this.get("CosInfos");
        return list && list.length;
    };

    Customer.fetch = function (c_num, options) {
        $.ajax({
            url: App.getApiPath("Customers"),
            data: { id: c_num },
            type: "GET",
            cache: false
        }).done(function(result){
            var c = new Customer(result);
            c.parseJSON();
            c.dirty = false;
            if (options && options.success) options.success(c);
        }).fail(function (jqXHR, textStatus, errorThrown) {
            if (options && options.fail) options.fail(jqXHR, textStatus, errorThrown);
        }).always(function () {
            if (options && options.always) options.always();
        });
    };

    Customer.updateFinalInfo = function (c_num, finalize, options) {
        $.ajax({
            url: App.getApiPath("customers/" + c_num + "/finalize/" + finalize),
            type: "POST",
            processData: false,
            contentType: "application/json; charset=utf-8"
        })
        .done(function (result) {
            if (options && options.success) options.success(result);
        }).fail(function (jqXHR, textStatus, errorThrown) {
            if (options && options.fail) options.fail(jqXHR, textStatus, errorThrown);
        }).always(function () {
            if (options && options.always) options.always();
        });
    };

    Customer.saveNote = function (c_num, note, edit_all, options) {
        var data = {
            c_num: c_num,
            note: note,
            edit_all: edit_all
        };
        var dataStr = JSON.stringify(data);
        //console.log("Customer.save: dataStr = " + dataStr);

        $.ajax({
            url: App.getApiPath("customers/note"),
            data: dataStr,
            type: "POST",
            processData: false,
            contentType: "application/json; charset=utf-8"
        })
            .done(function (result) {
                if (options && options.success) options.success(result);
            }).fail(function (jqXHR, textStatus, errorThrown) {
                if (options && options.fail) options.fail(jqXHR, textStatus, errorThrown);
            }).always(function () {
                if (options && options.always) options.always();
            });
    };

    return Customer;
});